x-pack/platform/plugins/shared/fleet/server/tasks/upgrade_agentless_deployment.ts

```diff
@@ -0,0 +1,276 @@
+    }
+
+    this.startedTaskRunner = true;
+    this.logger.info(`${LOGGER_SUBJECT} Started with interval of [${INTERVAL}]`);
+
+    try {
+      await taskManager.ensureScheduled({
+        id: this.taskId,
+        taskType: UPGRADE_AGENTLESS_DEPLOYMENTS_TASK_TYPE,
+        schedule: {
+          interval: INTERVAL,
+        },
+        state: {},
+        params: { version: UPGRADE_AGENT_DEPLOYMENTS_TASK_VERSION },
+      });
+    } catch (e) {
+      this.logger.error(`Error scheduling task ${LOGGER_SUBJECT}, error: ${e.message}`, e);
+    }
+  };
+
+  private get taskId(): string {
+    return `${UPGRADE_AGENTLESS_DEPLOYMENTS_TASK_TYPE}:${UPGRADE_AGENT_DEPLOYMENTS_TASK_VERSION}`;
+  }
+
+  private endRun(msg: string = '') {
+    this.logger.info(`${LOGGER_SUBJECT} runTask ended${msg ? ': ' + msg : ''}`);
+  }
+
+  private processInBatches = async (
+    {
+      agentlessPolicies,
+      agents,
+    }: {
+      agentlessPolicies: AgentPolicy[];
+      agents: Agent[];
+    },
+    batchSize: number,
+    processFunction: (agentPolicy: AgentPolicy, agentlessAgent: Agent) => void
+  ) => {
+    for (let i = 0; i < agentlessPolicies.length; i += batchSize) {
+      const currentAgentPolicyBatch = agentlessPolicies.slice(i, i + batchSize);
+
+      await Promise.allSettled(
+        currentAgentPolicyBatch.map((agentPolicy) => {
+          const agentlessAgent = agents.find((agent) => agent.policy_id === agentPolicy.id);
+
+          if (!agentlessAgent) {
+            this.endRun('No active online agentless agent found');
+            return;
+          }
+
+          this.logger.info(
+            `${LOGGER_SUBJECT} processing agentless agent ${JSON.stringify(agentlessAgent.agent)}`
+          );
+
+          return processFunction(agentPolicy, agentlessAgent);
+        })
+      );
+    }
+  };
+
+  private processUpgradeAgentlessDeployments = async (
+    esClient: ElasticsearchClient,
+    soClient: SavedObjectsClient
+  ) => {
+    const SAVED_OBJECT_TYPE = 'fleet-agent-policies';
+
+    const policiesKuery = `${SAVED_OBJECT_TYPE}.supports_agentless: true`;
+
+    try {
+      const agentPolicyFetcher = await agentPolicyService.fetchAllAgentPolicies(soClient, {
+        kuery: policiesKuery,
+        perPage: BATCH_SIZE,
+      });
+      this.logger.info(
+        `[${LOGGER_SUBJECT}] running task to upgrade agentless deployments with kuery: ${policiesKuery}`
+      );
+
+      for await (const agentPolicyPageResults of agentPolicyFetcher) {
+        this.logger.debug(
+          `[${LOGGER_SUBJECT}] Found "${agentPolicyPageResults.length}" agentless policies`
+        );
+
+        if (!agentPolicyPageResults.length) {
+          this.endRun('Found no agentless policies to process');
+          return;
+        }
+
+        const agentlessPolicies = agentPolicyPageResults.filter(async (agentPolicy) => {
+          return !!agentPolicy.package_policies?.length && !!agentPolicy.agents;
```

Comment:
I think `package_policies` and `agents` are never populated from `fetchAllAgentPolicies`

Reply:
@nchaulet  Just curious why will the `package_policies` and `agents` never be populated?

Reply:
The `package_policies` fields are populated only when retrieving a unique policy with `agentPolicyService.get` or when passing the options with `withPackagePolicies` and
`withAgentCount`  `agentPolicyService.list` as it makes some extra calls to populate those fields, we should probably improve our types in the future to reflect that.